pkg/resolution/resource/crd_resource_test.go

```diff
@@ -0,0 +1,260 @@
+		expectedErr              error
+	}{
+		{
+			name:                     "resolution request does not exist and needs to be created",
+			inputRequest:             nil,
+			expectedRequest:          getResolutionRequestFromFile("testdata/submit.resolution_request.created.yaml"),
+			expectedResolvedResource: nil,
+			expectedErr:              resolutioncommon.ErrRequestInProgress,
+		},
+		{
+			name:                     "resolution request exist and status is unknown",
+			inputRequest:             getResolutionRequestFromFile("testdata/submit.resolution_request.unknown.yaml"),
+			expectedRequest:          nil,
+			expectedResolvedResource: nil,
+			expectedErr:              resolutioncommon.ErrRequestInProgress,
+		},
+		{
+			name:                     "resolution request exist and status is true",
+			inputRequest:             getResolutionRequestFromFile("testdata/submit.resolution_request.true.yaml"),
+			expectedRequest:          nil,
+			expectedResolvedResource: &readOnlyResolutionRequest{getResolutionRequestFromFile("testdata/submit.resolution_request.true.yaml")},
+			expectedErr:              nil,
+		},
+		{
+			name:                     "resolution request exist and status is false",
+			inputRequest:             getResolutionRequestFromFile("testdata/submit.resolution_request.false.yaml"),
+			expectedRequest:          nil,
+			expectedResolvedResource: nil,
+			expectedErr:              resolutioncommon.NewError(resolutioncommon.ReasonResolutionFailed, errors.New("message")),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			d := test.Data{}
+			if tc.inputRequest != nil {
+				d.ResolutionRequests = []*v1beta1.ResolutionRequest{tc.inputRequest}
+			}
+
+			testAssets, cancel := getCRDRequester(t, d)
+			defer cancel()
+			ctx := testAssets.Ctx
+			clients := testAssets.Clients
+
+			resolver := ResolverName("git")
+			request := getRequestFromFile("testdata/request.golden.yaml")
+			crdRequester := NewCRDRequester(clients.ResolutionRequests, testAssets.Informers.ResolutionRequest.Lister())
+			resolvedResource, err := crdRequester.Submit(ctx, resolver, request.Request())
+
+			// check the error
+			if (tc.expectedErr != nil && err == nil) ||
+				(tc.expectedErr == nil && err != nil) {
+				t.Fatalf("expected error %v, but got %v", tc.expectedErr, err)
+			} else if tc.expectedErr != nil {
+				if d := cmp.Diff(tc.expectedErr.Error(), err.Error()); d != "" {
+					t.Fatalf("expected error to match %s", diff.PrintWantGot(d))
+				}
+			}
+
+			// check the resolved resource
+			if (tc.expectedResolvedResource != nil && resolvedResource == nil) ||
+				(tc.expectedResolvedResource == nil && resolvedResource != nil) {
+				t.Fatalf("expected resolved resource equal %v, but got %v", tc.expectedResolvedResource, resolvedResource)
+			} else if tc.expectedResolvedResource != nil {
+				if d := cmp.Diff(tc.expectedResolvedResource.req, resolvedResource.(readOnlyResolutionRequest).req); d != "" {
```

Comment:
Not working well.
I have tried to directly perform a diff, but it will panic.
```
panic: cannot handle unexported field at {*resource.readOnlyResolutionRequest}.req:
	"github.com/tektoncd/pipeline/pkg/resolution/resource".readOnlyResolutionRequest
consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported [recovered]
	panic: cannot handle unexported field at {*resource.readOnlyResolutionRequest}.req:
	"github.com/tektoncd/pipeline/pkg/resolution/resource".readOnlyResolutionRequest
consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported
```

Reply:
It looks like this test is trying to test things that aren't exported, i.e. not part of the API of this package. Maybe it would make sense to instead call `Annotations` and `Data`, which are exported, to test the parts of the package that its users will interact with? I think this is a reasonable exception to the code standard about testing only one function at a time.

Reply:
Your suggestion is excellent, I will make some adjustments.